scapy/utils.py

```diff
@@ -784,7 +784,11 @@ def read_packet(self, size=MTU):
         s,(sec,usec,wirelen) = rp
         
         try:
-            p = self.LLcls(s)
+            # Check for IPv6 when the linktype is RawIP
+            if self.linktype == DLT_RAW and ord(s[0]) >> 4 == 6:
```

Comment:
I really don't like to add those tests **for each packet** we read.

A good alternative solution would be to introduce a `IPv46` class which inherits from `IP` and uses a `dispatch_hook` to decide whether the IPv4 or IPv6 dissector should be used:
```diff
--- a/scapy/layers/inet6.py
+++ b/scapy/layers/inet6.py
@@ -3074,6 +3074,17 @@ _mip6_mhtype2cls = { 0: MIP6MH_BRR,
                      7: MIP6MH_BE }
 
 
+class _IPv46(IP):
+    @classmethod
+    def dispatch_hook(cls, _pkt=None, *_, **kargs):
+        if _pkt:
+            if struct.unpack('B', _pkt[0])[0] >> 4 == 6:
+                return IPv6
+        elif kargs.get('version') == 6:
+            return IPv6
+        return IP
+
+
 #############################################################################
 #############################################################################
 ###                             Traceroute6                               ###
@@ -3858,6 +3869,8 @@ def NDP_Attack_Fake_Router(ra, iface=None, mac_src_filter=None,
 
 conf.l3types.register(ETH_P_IPV6, IPv6)
 conf.l2types.register(31, IPv6)
+conf.l2types.register(DLT_IPV6, IPv6)
+conf.l2types.register(DLT_RAW, _IPv46)
 
 bind_layers(Ether,     IPv6,     type = 0x86dd )
 bind_layers(CookedLinux, IPv6,   proto = 0x86dd )
```

What do you think?